### ISSUE TITLE
Individuals are tracked by Google Analytics.

### ISSUE THREAD

**Issue by robdyke** [nhsx/COVID-19-app-Android-BETA#11]
**Describe the bug**

Individuals are tracked by Google Analytics.

When accessing the Privacy Policy tracking code is passed from the application to the covid19.nhs.uk website which is processed by Google Analytics. Data captured could be used to re-identify an individual.

https://github.com/nhsx/COVID-19-app-Android-BETA/blob/43a167f8dba422fd9001b64f9c4fd82275abb1c8/app/src/main/java/uk/nhs/nhsx/sonar/android/app/util/BrowserUtils.kt#L20-L25

**Expected behaviour**

Individuals are NOT tracked by Google Analytics

**Comment by technoir42** [comment]
Can you elaborate how that data can be used to identify an individual? Those UTM tags don't contain any PII.

**Comment by r3m0t** [comment]
@technoir42 is there going to be a privacy policy for the privacy policy? ;-)

**Comment by ShirtlessKirk** [comment]
> Those UTM tags don't contain any PII.

Because those are the app specific ones for use by the owners when looking at reports. Plenty of other information is sent along with IPs, OS, browser / webview, screen dimensions, "supercookie" user id references that link to every other Google service with analytics applied (including location), plus crosslinks to Facebook behind the scenes...

To think that all Google gets sent is the values of these constants linked to a GA ID and a timestamp seems willfully ignorant.

**Comment by technoir42** [comment]
@ShirtlessKirk I'm just trying to be reasonable. The app opens privacy policy in a custom tab, which is basically the regular Chrome. Google Analytics has access to cookies and whatever is accessible through your Chrome. If you the privacy policy yourself in the browser (or any other NHS web page) the same stuff will be sent to Google.
The app adds some extra tags to identify where the user came from but nothing that can be considered PII so I don't see anything wrong with the app.

**Comment by billysielu** [comment]
Whichever way this goes, Q22 here is relevant: 
https://committees.parliament.uk/oralevidence/334/html/

**Comment by talwrii** [comment]
@ShirtlessKirk 

Presumably the IP address and time of registration could be used to tie the user to the server side generated key. 

There's a general problem with even having to *talk* to a server.

**Comment by Xenoamor** [comment]
Yes, as you say @talwrii I don't think there is a way to reliably prevent this issue with a centralised system. A government could always log the IP of any and all connections and could request the information from your ISP or phone provider.

The only way around this I can think of is to use an anonymous network like Tor. Of course that has it's own challenges though

**Comment by djh42** [comment]
> If you open the same URL yourself in the browser (or any other NHS web page) the same stuff will be sent to Google.

No. If I open a page in a browser, noscript an ublock and privacybadger prevent anything being sent to google analytics (I hope :). In an app like this, I would expect it to be engineered not to send ad tracking data like this. If that means the web pages on the server need to be changed to not include the relevant code, so be it. (Note that I never access web pages from my phone in ordinary use - it's for making telephone calls and receiving SMS messages.

**Comment by hithomasmorelli** [comment]
Don't all the pages users are sent to have their own analytics consent box? I got the one below from visiting https://covid19.nhs.uk/privacy-and-data.html
![Consent box for analytics cookies, found at https://covid19.nhs.uk/privacy-and-data.html](https://user-images.githubusercontent.com/22722644/81472722-dd0d3e80-91f1-11ea-824d-a4378f3b3db7.png)

**Comment by Synchro** [comment]
PII is not a thing that exists under GDPR. The term you’re looking for is personal data, and all of this tracking qualifies.

**Comment by ripixelcodes** [comment]
Tracking via GA is only a breach of GDPR depending on how you've configured your GA. If you've disabled advertising features, and you've turned on anonymize IP, then no consent is required, and no "worrying" tracking is occurring (including indirect PII). Sharing their GA configuration would be a good gesture! As this app is open source we can see what data is being sent to GA, it seems to only be used for tracking what sections/pages of an app/site are being accessed.

Not all tracking is evil. Crash analytics, usage stats, and pain point identification are all expected and encouraged in good application development. I'm sure any "dodgy" usage will be spotted immediately, and will always be in the NHS' best interest to avoid. The whole purpose of these apps are public health, and thus confidence in them is second only to actual usefulness - something I'm sure NHSX are very aware of, and will do their best to maintain.

**Comment by Synchro** [comment]
For web use, that’s only true if consent is obtained _before_ loading GA’s client scripts; the anonymisation feature is cosmetic as the IP is leaked anyway. In an app it may be different as the client is embedded, reducing exposure somewhat.

**Comment by sftcd** [comment]
FYI, myself and a colleague Doug Leith (actually mostly him) took a look at opentrace (the open-source flavour of the Singapore app), and found that their use of Google Firebase would enable Google as well as the Singapore govt to track users of that app. We wrote up a [report](https://www.scss.tcd.ie/Doug.Leith/pubs/opentrace_privacy.pdf) on that.

From Doug's first look at this one, the UK app seems to have similar issues. When/if we bottom out on those, we'll likely write that up too, unless someone else has done it first.

**Comment by edent** [comment]
Thanks for all the comments. As [we say in the blog](https://www.nhsx.nhs.uk/blogs/code-behind-nhs-covid-19-app/)

> The public version of the app does not use any third-party analytics. In order to improve the app, we need to collect information about what happens if it crashes. This will use the operating system's default error-reporting tools.
> Our closed Beta will collect some volunteers' data for performance analytics and A/B testing. The libraries required for these analytics may still be present - but deactivated -  in the public version of the app.

I agree that we probably don't want UTM trackers on the links.

**Comment by kuro68k** [comment]
GDPR violation or not, this does not demonstrate good faith on the behalf of the NHS. If trust is to be built in a centralized model app merely reading the privacy policy should not result in analytics, or even a request to do analytics.

Trust in the app already seems to be low, especially due to the need for the location permission. This should be a fairly easy fix and worth doing just to built trust.

**Comment by Xenoamor** [comment]
Is there the option to disable crash reports or ask for the user's confirmation before sending one? Make/model/OS_version etc.. can all lead to de-anonymisation

**Comment by robdyke** [comment]
> I agree that we probably don't want UTM trackers on the links.

Would you accept a PR that removes these?

**Comment by edent** [comment]
Happy for you to send a PR.  That said, this repository represents the code of the *actual* running app. I don't want people being confused by seeing UTM in the app but not in the code.

I want us to move to working in the open, so that public pull requests can be merged into the live code (after suitable review). That still leaves us with the problem that you might be running v1.1 and the code on the website is v1.2.

**Comment by Synchro** [comment]
Surely you can make that clear by appropriate release tagging? Or by only accepting PRs into a dev branch that gets merged to master on release so that you can keep master reflecting live code?

**Comment by djh42** [comment]
> I want us to move to working in the open

@edent, I was going to ask about the back end, but I noticed you also say in the blog that you're going to open-source that too. Any idea when?

**Comment by markalanrichards** [comment]
> Tracking via GA is only a breach of GDPR (without getting consent) depending on how you've configured your GA. If you've disabled advertising features, and you've turned on anonymize IP, then no consent is required, and no "worrying" tracking is occurring (including indirect PII). Sharing their GA configuration would be a good gesture! As this app is open source we can see what data is being sent to GA, it seems to only be used for tracking what sections/pages of an app/site are being accessed.
> 
> Not all tracking is evil. Crash analytics, usage stats, and pain point identification are all expected and encouraged in good application development. I'm sure any "dodgy" usage will be spotted immediately, and will always be in the best interests of the NHS to avoid. The whole purpose of these apps are public health, and thus confidence in them is second only to actual usefulness - something I'm sure NHSX are very aware of, and will do their best to maintain.

But what about tracking regarding PECR (cookie law)? We have rights regarding electronic device privacy not just personal data. Unnecessary collection of data from our device by a third party, regardless of how identifiable it is, should not be happening.

Secretive crash analytics, usage stats and pain point gathering from client side is not good application development, it's immoral and potentially unlawful application development. Software existed for decades before the internet and still does on many non-internet connecting devices (especially in healthcare) with out needing to collect data without consent and many good websites and client side software that are connected to the internet continue in this spirit. I've worked on multiple applications that have demonstrated this respect. Good application development means providing a user interface for the user to be asked if they are happy to share data and being told what data is sent and ensuring it only goes to the necessary parties.

Regarding the evil nature, I would hope of a lot of heat in this debate is about securing trust.
For years, the NHS included advertiser tracking on its primary web sites (and in some places it still does). As an organisation it has shown a record of incredible incompetence at information security and privacy. This app is young and it has the potential to fall into the same pitfalls as too many other NHS software efforts.

**Comment by finlaand** [comment]
Option A, ask x million users to change dev options on their phone to install ring fenced apk, that still would require xyz settings changed based on iteration of the android api used... Car crash

Option B, land the app in play store (last I did you enabled tracking all over the place in a myriad of ways) whether wanted to or not, regardless of apis for fabric crash analytics, or Google analytics used).. In other words the real world and what happens with splunk behind the scenes and correlation of unstructured datasets

Option C.... Answer... If you are time pressed, just time bomb the app, literally, it stops working as of Jan 1 2021, forces future update or thoughts on how to resolve.

**Comment by philipswift** [comment]
I think that MAC addresses of device ports are used in pro-active monitoring as IP can change; even if the end-user has a fixed IP address with the ISP. Triangulation of mobile phone towers and demanding data from ISP's can only happen legally if an 'incident' has been logged with the authorities. Doing a quick mental cost/benefit analysis, all this is privacy banter is irrelevant if it a) saves life (no point monitoring/spying on a cadaver!) and b) considerably adds to a robust and sustainable knowledge base for coping with future viral outbreaks.

**Comment by talwrii** [comment]
> Doing a quick mental cost/benefit analysis, all this is privacy banter is irrelevant if it a) saves life (no point monitoring/spying on a cadaver!) and b) considerably adds to a robust and sustainable knowledge base for coping with future viral outbreaks.

Ugh, that's statement is more inflammatory than you might imagine:

* Lots of people mightn't consider this "privacy banter" but legitimate concern for an app that is going to be installed on a mass scale.
* "a) saves life (no point monitoring/spying on a cadaver!)"  there is almost always a trade off between safety and other concerns. For example, we are allowed to drive despite the fact that it kills thousands of people a year, thousands of people have died for political freedom. The question is what the trade is in this casen.
* "b) considerably adds to a robust and sustainable knowledge base for coping with future viral outbreaks." the same logic applies as for a.

**Comment by kuro68k** [comment]
Have you looked at what the police actually have to do to request that kind of data? It's basically self service for them, and they sign off hundreds of thousands of requests a year for the Met alone.

Basically if it is collected then the police have it, no question.

**Comment by PrivacyMatters** [comment]
> Can you elaborate how that data can be used to identify an individual? Those UTM tags don't contain any PII.

In the context of online tracking EU ePrivacy law applies irrespective of whether information is personal data or not. EU law is based on the concept of personal data rather than PII.  If you wish to access or store information on a person's terminal equipment, then you need a person's consent unless access / storage is strictly necessary for a service requested by an individual.

**Comment by philipswift** [comment]
Semantics. Should, could, would... Best fit solutions, best practise and continuous improvement (with KPI's) are the main drivers. The 3 worlds **we all** live in **at the same time** (not Quantum either) are 1) what we actually do 2) what is written down that we must/should do and 3) what 'the powers that be' said we did/do.  Laws change like belief systems. GDPR and PECR are under UK Government Governance and as the Government is pushing this App the GDPR and PECR laws and governance are negated in real-world terms. Officially and idealistically they remain written in stone as a statute. 
![image](https://user-images.githubusercontent.com/6306791/82200901-8097d680-98f7-11ea-959c-b667c8cfb85d.png)
Yeah yeah....I'm getting off it.

**Comment by edent** [comment]
I'm pasting this message in every active GitHub issue, so you may receive duplicate notifications.

Today, I'm happy to announce that NHSX has released the full git commit history for the Isle of Wight Beta apps.

* Android - https://github.com/nhsx/COVID-19-app-Android-BETA
* iOS - https://github.com/nhsx/COVID-19-app-iOS-BETA

As discussed, we have redacted API keys, sensitive domain names, and some of the developers' personal details. I am still waiting on final approval to publish the server-side code.

I would like to personally thank the community for your comments, bug reports, and vulnerability disclosures. They all went into helping the development process.

The beta trial of this app has now ended and we've moved to the next phase of app development. It is our intention to publish the source code of future apps as the binaries are released to the public.

Once again, thank you for being part of this.

Terence Eden
Head of Open Technology - NHSX
